dvc/render/data.py

```diff
@@ -0,0 +1,211 @@
+from copy import deepcopy
+from functools import partial
+from typing import Dict, List, Optional, Set, Union
+
+from funcy import first, reraise
+
+from dvc.exceptions import DvcException
+from dvc.render.base import INDEX_FIELD, REVISION_FIELD
+
+
+class FieldsNotFoundError(DvcException):
+    def __init__(self, expected_fields, found_fields):
+        self.expected_fields = expected_fields
+        self.found_fields = found_fields
+
+        expected_str = ", ".join(expected_fields)
+        found_str = ", ".join(found_fields)
+        super().__init__(
+            f"Could not find all provided fields ('{expected_str}') "
+            f"in data fields ('{found_str}')."
+        )
+
+
+class PlotDataStructureError(DvcException):
+    def __init__(self):
+        super().__init__(
+            "Plot data extraction failed. Please see "
+            "https://man.dvc.org/plots for supported data formats."
+        )
+
+
+def _filter_fields(datapoints: List[Dict], fields: Set) -> List[Dict]:
+    if not fields:
+        return datapoints
+    assert isinstance(fields, set)
+
+    new_data = []
+    for data_point in datapoints:
+        keys = set(data_point.keys())
+        if keys & fields != fields:
+            raise FieldsNotFoundError(fields, keys)
+
+        new_data.append(
+            {key: value for key, value in data_point.items() if key in fields}
+        )
+
+    return new_data
+
+
+def _lists(dictionary: Dict):
+    for _, value in dictionary.items():
+        if isinstance(value, dict):
+            yield from _lists(value)
+        elif isinstance(value, list):
+            yield value
+
+
+def _find_first_list(data: Union[Dict, List], fields: Set) -> List[Dict]:
```

Comment:
I'm sorry I'm not quite familiar with the vega render but looks like this function is not deterministic, without
 any problem?

Reply:
What do you mean by not deterministic? Are you talking about `_lists` sourcing from `Dict`? AFAIK, in python3 the Dicts are ordered.

Reply:
It will give a different result even if we put the same arguments to it?

Reply:
I am not sure what is the situation when it would return different results? If you are talking about the dictionary and the `_lists` call, python dicts maintain the [order](https://stackoverflow.com/questions/39980323/are-dictionaries-ordered-in-python-3-6). We had few discussions (eg https://github.com/iterative/dvc/issues/5009) and AFAIR we decided to not put too much effort in asserting the order of dict "just in case", but react if default behavior will change.